File: pkg/models/wiki.go
package models

import (
	"time"
)

// Wiki represents a generated wiki
type Wiki struct {
	ID           string                `json:"id"`
	RepositoryID string                `json:"repository_id"`
	PackagePath  string                `json:"package_path"` // e.g., github.com/gorilla/websocket
	Title        string                `json:"title"`
	Description  string                `json:"description"`
	Status       WikiStatus            `json:"status"`
	Progress     int                   `json:"progress"` // 0-100
	Tags         []string              `json:"tags"`     // Wiki-level tags
	CreatedAt    time.Time             `json:"created_at"`
	UpdatedAt    time.Time             `json:"updated_at"`
	GeneratedBy  string                `json:"generated_by"` // AI provider used
	Model        string                `json:"model"`        // AI model used
	Language     string                `json:"language"`     // Primary language of this wiki
	Languages    []string              `json:"languages"`    // All available languages
	Pages        []WikiPage            `json:"pages"`
	Diagrams     []WikiDiagram         `json:"diagrams"`
	Settings     WikiSettings          `json:"settings"`
	Metadata     WikiMetadata          `json:"metadata"`
	Translations map[string]*WikiTrans `json:"translations,omitempty"` // Language code -> translation
}

// WikiStatus represents the status of wiki generation
type WikiStatus string

const (
	WikiStatusPending    WikiStatus = "pending"
	WikiStatusAnalyzing  WikiStatus = "analyzing"
	WikiStatusGenerating WikiStatus = "generating"
	WikiStatusCompleted  WikiStatus = "completed"
	WikiStatusFailed     WikiStatus = "failed"
)

// LogLevel represents the level of a log entry
type LogLevel string

const (
	LogLevelInfo    LogLevel = "info"
	LogLevelWarning LogLevel = "warning"
	LogLevelError   LogLevel = "error"
	LogLevelSuccess LogLevel = "success"
	LogLevelDebug   LogLevel = "debug"
)

// WikiTrans represents a translation of a wiki
type WikiTrans struct {
	Language    string        `json:"language"`    // Language code (e.g., "en", "zh", "ja")
	Title       string        `json:"title"`       // Translated title
	Description string        `json:"description"` // Translated description
	Pages       []WikiPage    `json:"pages"`       // Translated pages
	Diagrams    []WikiDiagram `json:"diagrams"`    // Translated diagrams
	CreatedAt   time.Time     `json:"created_at"`
	UpdatedAt   time.Time     `json:"updated_at"`
}

// SupportedLanguages defines the supported languages for wiki generation
var SupportedLanguages = map[string]string{
	"en": "English",
	"zh": "中文",
	"ja": "日本語",
	"ko": "한국어",
	"es": "Español",
	"fr": "Français",
	"de": "Deutsch",
	"ru": "Русский",
	"pt": "Português",
	"it": "Italiano",
}

// WikiLogEntry represents a detailed log entry
type WikiLogEntry struct {
	Timestamp  time.Time `json:"timestamp"`
	Level      LogLevel  `json:"level"`
	Step       string    `json:"step"`
	Message    string    `json:"message"`
	Details    string    `json:"details,omitempty"`
	Duration   string    `json:"duration,omitempty"`
	Progress   int       `json:"progress,omitempty"`
	FilesCount int       `json:"files_count,omitempty"`
	TokensUsed int       `json:"tokens_used,omitempty"`
	Error      string    `json:"error,omitempty"`
}

// WikiPage represents a page in the wiki
type WikiPage struct {
	ID          string    `json:"id"`
	Title       string    `json:"title"`
	Content     string    `json:"content"`
	Type        PageType  `json:"type"`
	Order       int       `json:"order"`
	ParentID    string    `json:"parent_id,omitempty"`
	Children    []string  `json:"children,omitempty"`
	Tags        []string  `json:"tags,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	WordCount   int       `json:"word_count"`
	ReadingTime int       `json:"reading_time"` // in minutes
}

// PageType represents the type of wiki page
type PageType string

const (
	PageTypeOverview     PageType = "overview"
	PageTypeArchitecture PageType = "architecture"
	PageTypeAPI          PageType = "api"
	PageTypeModule       PageType = "module"
	PageTypeFunction     PageType = "function"
	PageTypeClass        PageType = "class"
	PageTypeTutorial     PageType = "tutorial"
	PageTypeReference    PageType = "reference"
	PageTypeChangelog    PageType = "changelog"
	PageTypeGuide        PageType = "guide"
)

// WikiDiagram represents a diagram in the wiki
type WikiDiagram struct {
	ID          string      `json:"id"`
	Title       string      `json:"title"`
	Type        DiagramType `json:"type"`
	Content     string      `json:"content"` // Mermaid syntax
	Description string      `json:"description"`
	PageID      string      `json:"page_id,omitempty"`
	CreatedAt   time.Time   `json:"created_at"`
	UpdatedAt   time.Time   `json:"updated_at"`
}

// DiagramType represents the type of diagram
type DiagramType string

const (
	DiagramTypeFlowchart    DiagramType = "flowchart"
	DiagramTypeSequence     DiagramType = "sequence"
	DiagramTypeClass        DiagramType = "class"
	DiagramTypeER           DiagramType = "er"
	DiagramTypeGantt        DiagramType = "gantt"
	DiagramTypeGitGraph     DiagramType = "gitgraph"
	DiagramTypeArchitecture DiagramType = "architecture"
	DiagramTypeDataFlow     DiagramType = "dataflow"
)

// WikiSettings represents wiki generation settings
type WikiSettings struct {
	AIProvider      string            `json:"ai_provider"`
	Model           string            `json:"model"`
	Temperature     float32           `json:"temperature"`
	MaxTokens       int               `json:"max_tokens"`
	EnableDiagrams  bool              `json:"enable_diagrams"`
	EnableRAG       bool              `json:"enable_rag"`
	Language        string            `json:"language"`
	Theme           string            `json:"theme"`
	CustomPrompts   map[string]string `json:"custom_prompts,omitempty"`
	ExcludePatterns []string          `json:"exclude_patterns,omitempty"`
	IncludePatterns []string          `json:"include_patterns,omitempty"`
}

// WikiMetadata represents additional metadata about the wiki
type WikiMetadata struct {
	GenerationTime    time.Duration  `json:"generation_time"`
	TokensUsed        int            `json:"tokens_used"`
	FilesProcessed    int            `json:"files_processed"`
	PagesGenerated    int            `json:"pages_generated"`
	DiagramsGenerated int            `json:"diagrams_generated"`
	Languages         []string       `json:"languages"`
	Complexity        string         `json:"complexity"` // low, medium, high
	Quality           float64        `json:"quality"`    // 0-1
	Tags              []string       `json:"tags"`
	Categories        []string       `json:"categories"`
	Statistics        map[string]int `json:"statistics"`
	PackagePath       string         `json:"package_path"`   // 包路径，如 github.com/gin-gonic/gin
	RepositoryURL     string         `json:"repository_url"` // 原始仓库URL
}

// GenerationRequest represents a request to generate a wiki
type GenerationRequest struct {
	RepositoryURL    string       `json:"repository_url"`
	Branch           string       `json:"branch,omitempty"`
	AccessToken      string       `json:"access_token,omitempty"`
	Settings         WikiSettings `json:"settings"`
	Title            string       `json:"title,omitempty"`
	Description      string       `json:"description,omitempty"`
	CustomPrompts    []string     `json:"custom_prompts,omitempty"`
	Languages        []string     `json:"languages,omitempty"`          // Languages to generate (e.g., ["en", "zh"])
	PrimaryLanguage  string       `json:"primary_language,omitempty"`   // Primary language (default: "en")
	GenerateAllLangs bool         `json:"generate_all_langs,omitempty"` // Generate all supported languages
}

// GenerationProgress represents the progress of wiki generation
type GenerationProgress struct {
	WikiID      string     `json:"wiki_id"`
	Status      WikiStatus `json:"status"`
	Progress    int        `json:"progress"`
	CurrentStep string     `json:"current_step"`
	Message     string     `json:"message"`
	Error       string     `json:"error,omitempty"`
	StartedAt   time.Time  `json:"started_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	ETA         *time.Time `json:"eta,omitempty"`
}

// ChatMessage represents a message in the RAG chat system
type ChatMessage struct {
	ID         string      `json:"id"`
	WikiID     string      `json:"wiki_id"`
	Role       MessageRole `json:"role"`
	Content    string      `json:"content"`
	Context    []string    `json:"context,omitempty"` // Retrieved context
	Sources    []string    `json:"sources,omitempty"` // Source files
	Timestamp  time.Time   `json:"timestamp"`
	TokensUsed int         `json:"tokens_used,omitempty"`
}

// MessageRole represents the role of a chat message
type MessageRole string

const (
	MessageRoleUser      MessageRole = "user"
	MessageRoleAssistant MessageRole = "assistant"
	MessageRoleSystem    MessageRole = "system"
)

// SearchResult represents a search result in the wiki
type SearchResult struct {
	Type       string   `json:"type"` // page, diagram, code
	ID         string   `json:"id"`
	Title      string   `json:"title"`
	Content    string   `json:"content"`
	Excerpt    string   `json:"excerpt"`
	Score      float64  `json:"score"`
	Highlights []string `json:"highlights"`
	URL        string   `json:"url"`
	PageID     string   `json:"page_id,omitempty"`
	DiagramID  string   `json:"diagram_id,omitempty"`
}

// WikiExport represents exported wiki data
type WikiExport struct {
	Format    ExportFormat `json:"format"`
	Content   string       `json:"content"`
	Filename  string       `json:"filename"`
	Size      int64        `json:"size"`
	CreatedAt time.Time    `json:"created_at"`
}

// ExportFormat represents the format for wiki export
type ExportFormat string

const (
	ExportFormatMarkdown ExportFormat = "markdown"
	ExportFormatHTML     ExportFormat = "html"
	ExportFormatPDF      ExportFormat = "pdf"
	ExportFormatJSON     ExportFormat = "json"
	ExportFormatZIP      ExportFormat = "zip"
)

File: NewLife.Wiki/Models/Wiki.cs
using System;
using System.Collections.Generic;

namespace NewLife.Wiki.Models;

/// <summary>生成的 Wiki 实体，包含页面、图表及生成元数据。</summary>
public class Wiki
{
    /// <summary>Wiki 标识</summary>
    public String Id { get; set; } = String.Empty;
    /// <summary>关联的仓库 Id</summary>
    public String RepositoryId { get; set; } = String.Empty;
    /// <summary>打包输出路径</summary>
    public String PackagePath { get; set; } = String.Empty;
    /// <summary>标题</summary>
    public String Title { get; set; } = String.Empty;
    /// <summary>描述</summary>
    public String Description { get; set; } = String.Empty;
    /// <summary>当前状态</summary>
    public WikiStatus Status { get; set; } = WikiStatus.Pending;
    /// <summary>进度（0-100）</summary>
    public Int32 Progress { get; set; }
    /// <summary>标签集合</summary>
    public List<String> Tags { get; set; } = new();
    /// <summary>创建时间 UTC</summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>更新时间 UTC</summary>
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>生成工具或用户</summary>
    public String GeneratedBy { get; set; } = String.Empty;
    /// <summary>所用模型名称</summary>
    public String Model { get; set; } = String.Empty;
    /// <summary>主语言（默认 zh）</summary>
    public String Language { get; set; } = "zh";
    /// <summary>启用的语言列表</summary>
    public List<String> Languages { get; set; } = new();
    /// <summary>页面集合</summary>
    public List<WikiPage> Pages { get; set; } = new();
    /// <summary>图表集合</summary>
    public List<WikiDiagram> Diagrams { get; set; } = new();
    /// <summary>生成设置</summary>
    public WikiSettings Settings { get; set; } = new();
    /// <summary>元数据统计</summary>
    public WikiMetadata Metadata { get; set; } = new();
}

/// <summary>Wiki 生成阶段状态</summary>
public enum WikiStatus
{
    /// <summary>等待开始</summary>
    Pending,
    /// <summary>仓库分析中</summary>
    Analyzing,
    /// <summary>内容生成中</summary>
    Generating,
    /// <summary>完成</summary>
    Completed,
    /// <summary>失败</summary>
    Failed
}

/// <summary>页面类型</summary>
public enum PageType
{
    /// <summary>概览</summary>
    Overview,
    /// <summary>架构</summary>
    Architecture,
    /// <summary>API 参考</summary>
    Api,
    /// <summary>模块</summary>
    Module,
    /// <summary>函数</summary>
    Function,
    /// <summary>类</summary>
    Class,
    /// <summary>教程</summary>
    Tutorial,
    /// <summary>参考/索引</summary>
    Reference,
    /// <summary>变更日志</summary>
    Changelog,
    /// <summary>指南/使用</summary>
    Guide
}

/// <summary>图表类型</summary>
public enum DiagramType
{
    /// <summary>流程图</summary>
    Flowchart,
    /// <summary>时序图</summary>
    Sequence,
    /// <summary>类图</summary>
    Class,
    /// <summary>实体关系图</summary>
    ER,
    /// <summary>甘特图</summary>
    Gantt,
    /// <summary>Git 分支</summary>
    GitGraph,
    /// <summary>架构图</summary>
    Architecture,
    /// <summary>数据流图</summary>
    DataFlow
}

/// <summary>Wiki 单个页面</summary>
public class WikiPage
{
    /// <summary>页面 Id</summary>
    public String Id { get; set; } = String.Empty;
    /// <summary>标题</summary>
    public String Title { get; set; } = String.Empty;
    /// <summary>Markdown 内容</summary>
    public String Content { get; set; } = String.Empty;
    /// <summary>页面类型</summary>
    public PageType Type { get; set; }
    /// <summary>排序序号</summary>
    public Int32 Order { get; set; }
    /// <summary>父页面 Id</summary>
    public String? ParentId { get; set; }
    /// <summary>子页面 Id 列表</summary>
    public List<String> Children { get; set; } = new();
    /// <summary>标签</summary>
    public List<String> Tags { get; set; } = new();
    /// <summary>创建时间</summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>更新时间</summary>
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>字数统计</summary>
    public Int32 WordCount { get; set; }
    /// <summary>预计阅读时间（分钟）</summary>
    public Int32 ReadingTime { get; set; }
}

/// <summary>图表定义</summary>
public class WikiDiagram
{
    /// <summary>图表 Id</summary>
    public String Id { get; set; } = String.Empty;
    /// <summary>标题</summary>
    public String Title { get; set; } = String.Empty;
    /// <summary>类型</summary>
    public DiagramType Type { get; set; }
    /// <summary>源码（Mermaid/PlantUML 等）</summary>
    public String Content { get; set; } = String.Empty;
    /// <summary>描述说明</summary>
    public String Description { get; set; } = String.Empty;
    /// <summary>关联页面 Id</summary>
    public String? PageId { get; set; }
    /// <summary>创建时间</summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>更新时间</summary>
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>生成设置</summary>
public class WikiSettings
{
    /// <summary>AI 提供者标识</summary>
    public String AiProvider { get; set; } = String.Empty;
    /// <summary>模型名称</summary>
    public String Model { get; set; } = String.Empty;
    /// <summary>温度（0-1）</summary>
    public Single Temperature { get; set; } = 0.7f;
    /// <summary>最大生成 token 数</summary>
    public Int32 MaxTokens { get; set; } = 4000;
    /// <summary>是否生成图表</summary>
    public Boolean EnableDiagrams { get; set; } = true;
    /// <summary>是否启用 RAG（后续扩展）</summary>
    public Boolean EnableRag { get; set; } = true;
    /// <summary>主语言</summary>
    public String Language { get; set; } = "zh";
    /// <summary>主题</summary>
    public String Theme { get; set; } = "default";
    /// <summary>自定义提示词（键为用途，如 overview）</summary>
    public Dictionary<String,String> CustomPrompts { get; set; } = new();
    /// <summary>排除文件/目录模式</summary>
    public List<String> ExcludePatterns { get; set; } = new();
    /// <summary>包含文件/目录模式（为空表示全部）</summary>
    public List<String> IncludePatterns { get; set; } = new();
}

/// <summary>生成过程及结果统计</summary>
public class WikiMetadata
{
    /// <summary>生成耗时</summary>
    public TimeSpan GenerationTime { get; set; }
    /// <summary>使用 token 数</summary>
    public Int32 TokensUsed { get; set; }
    /// <summary>处理的文件数</summary>
    public Int32 FilesProcessed { get; set; }
    /// <summary>生成的页面数</summary>
    public Int32 PagesGenerated { get; set; }
    /// <summary>生成的图表数</summary>
    public Int32 DiagramsGenerated { get; set; }
    /// <summary>覆盖的语言列表</summary>
    public List<String> Languages { get; set; } = new();
    /// <summary>复杂度指标</summary>
    public String Complexity { get; set; } = String.Empty;
    /// <summary>质量评分（0~1 或百分比）</summary>
    public Double Quality { get; set; }
    /// <summary>标签合集</summary>
    public List<String> Tags { get; set; } = new();
    /// <summary>分类合集</summary>
    public List<String> Categories { get; set; } = new();
    /// <summary>统计字典，不同维度计数</summary>
    public Dictionary<String,Int32> Statistics { get; set; } = new();
    /// <summary>打包输出路径</summary>
    public String PackagePath { get; set; } = String.Empty;
    /// <summary>仓库地址</summary>
    public String RepositoryUrl { get; set; } = String.Empty;
}

/// <summary>Wiki 生成请求</summary>
public class GenerationRequest
{
    /// <summary>目标仓库地址</summary>
    public String RepositoryUrl { get; set; } = String.Empty;
    /// <summary>分支</summary>
    public String Branch { get; set; } = String.Empty;
    /// <summary>访问令牌（私有仓库可选）</summary>
    public String AccessToken { get; set; } = String.Empty;
    /// <summary>生成设置</summary>
    public WikiSettings Settings { get; set; } = new();
    /// <summary>自定义标题</summary>
    public String Title { get; set; } = String.Empty;
    /// <summary>自定义描述</summary>
    public String Description { get; set; } = String.Empty;
    /// <summary>额外自定义提示词（按顺序追加）</summary>
    public List<String> CustomPrompts { get; set; } = new();
    /// <summary>需生成的语言列表</summary>
    public List<String> Languages { get; set; } = new();
    /// <summary>主语言</summary>
    public String PrimaryLanguage { get; set; } = "zh";
    /// <summary>是否对全部检测到语言生成</summary>
    public Boolean GenerateAllLangs { get; set; }
}
